core/model-api/src/main/java/org/eclipse/rdf4j/model/Literal.java

```diff
@@ -290,4 +290,20 @@ default TemporalAmount temporalAmountValue() throws DateTimeException {
 	@Override
 	int hashCode();
 
+	/**
+	 * Checks whether the supplied value is a "plain literal". A "plain literal" is a literal with no datatype and
+	 * optionally a language tag.
+	 *
+	 * @see <a href="http://www.w3.org/TR/2004/REC-rdf-concepts-20040210/#dfn-plain-literal">RDF Literal
+	 *      Documentation</a>
+	 */
+	boolean isPlainLiteral();
+
+	/**
+	 * Checks whether the supplied value is a "simple literal". A "simple literal" is a literal with no language tag nor
+	 * datatype.
+	 *
+	 * @see <a href="http://www.w3.org/TR/sparql11-query/#simple_literal">SPARQL Simple Literal Documentation</a>
+	 */
+	boolean isSimpleLiteral();
```

Comment:
These two methods improve performance because there are many points where we can hard code the values or compute the values more efficiently for new instances.

Reply:
Plain literal as a concept doesn't really exist in RDF 1.1 anymore - a literal is either simple (no language tag), or if it has a language tag, it also has a datatype (and is therefore not plain). I wonder if we can't get rid of the logic surrounding that check completely. 

I'll try and find more time later this week to look into this PR in more depth. I think the idea looks good but I haven't fully grasped the consequences yet for compatibility, or for other store implementors. I also am a little concerned that having the logic about which literals plain/simple implemented in different places is going to cause inconsistencies (if I remember correctly this used to be handled by a utility class).

Reply:
Yeah. It used to be a utility class, which was really slow since it needed check for a language tag and try to match the datatype each time it was called, which is typically log(n) times for each literal when sorting.

Reply:
Oh yeah not saying this isn't a better approach (it is), just that I'd like to make sure we migrate properly and consistently.

Reply:
We could do a default implementation, but I would rather not. I don't really feel that these two methods need to be user facing, so I was thinking about making an ExtendedLiteral interface for them (and maybe other methods we want down the line). Then check for that interface with instanceof, and if not use the utility.

Though if you think we could drop one of the methods then maybe it's not too bad.

Reply:
If we merge https://github.com/eclipse/rdf4j/pull/3596 then these interface changes may not be needed. Enum based datatypes will make the helper utils much faster and probably fast enough.